### ISSUE TITLE
Alternate time-based timeline

### ISSUE THREAD

**Issue by kyrieru** [aseprite/aseprite#1342]
Somewhat of an odd request,

In games, sometimes you layer animations of different speeds. (imagine a 3 frame character attack and a 10 frame sword swing )

It would be useful if it were possible to play layers at different speeds somehow. Not to be exported that way, but to preview and draw them together.

The only problem is that I'm not so sure how this could be executed with the current timeline since it's based on linear frames as opposed to time, like in video editors. Only you would know if something like an alternate timeline is doable though.

**Comment by dacap** [comment]
Hi @kyrieru, actually this is a feature in my personal to-do list. I though there was already an issue about this, but I didn't find it.

The main idea is to create an alternative UI to manipulate the same internal [cel structures](https://www.aseprite.org/docs/cel/). For example, in this case:

![time-based-mockup](https://cloud.githubusercontent.com/assets/39654/20965095/0a50d0fa-bc53-11e6-90ab-23f9f08df82d.png)

We would have 6 frames with different durations and linked cels in the regular UI:

![cels-timeline](https://cloud.githubusercontent.com/assets/39654/20964950/7352ab74-bc52-11e6-9e34-3bfac3aa2fed.png)

So moving/changing the size of each "cel block" in the time-based timeline would change the duration of each "real cel" and create/delete linked cels/sections automatically.

**Comment by oceanhahn** [comment]
I'm really interested to see how this would look in action, too, given that each frame can have a unique duration. Do I see that represented in your sample on frames 2 (prolonged) and 4 (shortened), @dacap? :D

**Comment by dacap** [comment]
@oceanhahn yeah, and we can use linked cels to make frames even longer and combine them with other frames; the trick is to automatically adjust frames durations and create/delete linked cels each time we modify one frame-block in the time-based timeline.

**Comment by oceanhahn** [comment]
@dacap I can see how that would become particularly messy during export, too. :s 
I imagine that's a huge problem when more than two active layers are involved.

**Comment by dacap** [comment]
@oceanhahn I think the export process would be a different story, for example, when you use time-based timeline, is because: 1) you are creating just a gif-like animation, or 2) you want the time-based information in the JSON data.

The export to a sequence of `.png` files would be really messy (maybe forbidden).

**Comment by oceanhahn** [comment]
I suppose it's possible to do a png-sequence by superimposing a frame-based timeline (if the timeline architecture is revised to work on a duration basis) or exporting each frame, including the duration in its filename (_"swordswing0_10.48x63.140ms.png"_, for instance). It obviously won't produce a proper animation, but it would at least be easier for other software to interpret, potentially.

**Comment by kyrieru** [comment]
What about something like this?
![image](https://cloud.githubusercontent.com/assets/13920296/20990116/b787550a-bcad-11e6-85d1-59c77e81bdae.png)

As you scrub across, it would display the position of cells independent of each other, but the image displayed would be a combination of all of them. Then there would need to be a way of linking these smaller sections to other layers, tags, or the timeline, etc. The timing of the smaller section would be scaled to what it's linked to.

It's difficult to give specific practical suggestions because everything sounds so drastic. I'm used to talking to devs who aren't even willing to fix minor problems let alone make major changes, so I'm always afraid that suggestions might be drifting too far from your own goals for Aseprite. You really are exceptional in that regard.

Is there anything related to the timeline that you explicitly don't want to do, or any limitations that you would have to work with?

**Comment by dacap** [comment]
@kyrieru sorry, I'm a little confused by the image. What do represent those transparent rectangles? or the vertical lines?

> That said, it's difficult to know how practical of a suggestion any of this is.

You should think in terms of user-friendly and backward-compatibility always, the problem is that, as you said, we generally don't know how practical something will be until we used it. So sometimes the important thing is to create a prototype as fast as possible and discard it as fast as possible if it doesn't work.

Then internal details are mostly managed by me, and we can always find a way to adapt the code to something new. It doesn't matter how much time it will take as how much improvement it will make to the user daily experience.

> You really are exceptional in that regard, compared to other devs.

Thank you. I think it's because: 1) I'm working full-time on Aseprite so I have plenty of time to think about it, 2) I implement huge changes that only someone crazy like me would approve (e.g. [layers groups](https://github.com/aseprite/aseprite/issues/454), which was a massive number of changes).

> Is there anything related to the timeline that you explicitly don't want to do, or any limitations that you would have to work with?

No. Maybe I'd like to avoid #593 because the scope is too big (it affects almost every single aspect of the UI/sprite editor/timeline/etc.) and there are a lot of other issues before that one. But I think I'd finally implement that one too (as it's quite related to/useful for #532).

Anyway I'll finally filter or give more priority to things depending on its user-friendliness or the number of people requesting it, (Sometimes I change priorities of features just to avoid getting more user requests about it, as it's easier to program something than to give support about the non-existence of that something 😄 )

Finally we should always think on [progressive disclosure](https://en.wikipedia.org/wiki/Progressive_disclosure), so the program doesn't start getting bloated.

**Comment by kyrieru** [comment]
@dacap Aight, thanks for the clarification.

In the example I posted, the transparent rectangles and lines represent the progress of the animation. It's the equivalent of this-
https://i.gyazo.com/cd39c69ed58af41fddbe3d8cb87fae0c.png

- but for each layer, interdependently. My point with it was to show what it would look like if one layer were to play at a different speed than another. Clicking on a position on one layer would show the equivalent position on other layers, based on the speed/number of frames.

It would probably have to be a different kind of layer, with per-cell speeds. By default it would scale with the number of frames in the animation, but could also be linked to a tag, and would play during that tag instead. (In other words, it would show up on the timeline, but wouldn't behave the same way as other layers.)

**Comment by kyrieru** [comment]
Though, that doesn't solve the problem of frame-speed.

Another way would be to give all layers cell-based speed, and a checkbox that says "lock timing". When locked, layers would simply run at the speed of the frames, and when unlocked, they would play independent of it, like in the first picture I posted.

**Comment by dacap** [comment]
I'm going to try to implement the time-based timeline. Anyway the independent animation for each layer is more related to #593

**Comment by Doughnutsu** [comment]
> 
> 
> I'm going to try to implement the time-based timeline. Anyway the independent animation for each layer is more related to #593

Any chance this will come with the v1.3 update? I paid a talented person to make a few things and they did wonderfully using Aseprite. Though I've hit an issue now that the timings on the animations are different due their nature and can no longer be combined into one piece without messing up the timing.
